src/libmongoc/src/mongoc/mongoc-socket.c

```diff
@@ -782,6 +782,108 @@ mongoc_socket_bind (mongoc_socket_t *sock,       /* IN */
 }
 
 
+mongoc_socket_t *
+mongoc_socket_new_bind_then_connect (int domain,
+				     int type,
+				     int protocol,
+				     const char *bind_ip,
+				     const struct sockaddr *connect_addr,
+				     mongoc_socklen_t connect_addrlen,
+				     int64_t expire_at)
+{
+   mongoc_socket_t *sock;
+   struct sockaddr_in bind_addr;
+   int reuse = 1;
+   int ret;
+
+   ENTRY;
+
+   BSON_ASSERT (connect_addr);
+   BSON_ASSERT (connect_addrlen);
+
+   bind_addr.sin_port = htons (0);
+   
+   if (inet_pton (AF_INET, bind_ip, &bind_addr.sin_addr) == 1) {
+      bind_addr.sin_family = AF_INET;
+      if (domain != AF_INET && domain != PF_INET) {
+	 MONGOC_WARNING ("Cannot use an IPv4 bind IP without an IPv4 host");
+	 //return NULL;
+      }
+   } else if (inet_pton (AF_INET6, bind_ip, &bind_addr.sin_addr) == 1) {
+      bind_addr.sin_family = AF_INET6;
+      if (domain != AF_INET6 && domain != PF_INET6) {
+	 MONGOC_WARNING ("Cannot use an IPv6 bind IP without an IPv6 host");
+	 //return NULL;
+      }
+   } else {
+      MONGOC_WARNING ("Invalid bind address: %s", bind_ip);
+      return NULL;
+   }
+
+   fprintf (stderr, "passed-in family/domain is %d\n", domain);
+   fprintf (stderr, "determined family is %d\n", bind_addr.sin_family);
+
+   // TODO SAM: this loop shouldn't be infinite. Use expire_at ? Num retries? One retry?
```

Comment:
Thoughts on how to structure this loop?

This article talks about `SO_REUSEADDR` and explains why this retry loop is here:
https://idea.popcount.org/2014-04-03-bind-before-connect/

It is probably not strictly necessary to use `SO_REUSEADDR`, but I think it's a little more resilient to use it.  In a situation where the user's machine was opening a lot (like, a _lot_) of connections, including many made through the C driver, `SO_REUSEADDR` would help push up the hard limit on the number of connections that could be made.  But if we don't think it's worth using, then the logic in this method would be much simpler.

Reply:
Oh interesting, thanks for linking to that article.

> It is probably not strictly necessary to use SO_REUSEADDR, but I think it's a little more resilient to use it. 

I agree. From what I understand, closed sockets can be in the `TIME_WAIT` state for quite some time before they are eligible for reuse if `SO_REUSEADDR` is not set. It seems like 15 seconds on macOS `sysctl net.inet.tcp.msl`

If `SO_REUSEADDR` is set, it might be rare a user hits that limit. Though you make a good point that they may be creating sockets outside of the C driver. Even if binding to an IP, and using `SO_REUSEADDR` gives you 64k unique sockets to a destination IP, another application may have exhausted many of those sockets outside of the use of the C driver.

Waiting until `expire_at` seems reasonable. I am not sure how we could avoid spinning, if all sockets are exhausted, for, say 5 seconds, I'm not sure if it would loop at full CPU usage.

What if we don't retry? We don't know how this will be used. We can investigate more complex retry logic if we discover there is a need for it.

Reply:
I think that if we are going to set `SO_REUSEADDR`, we should retry at least once.  Because we are forcing the OS to pick an ephemeral socket for us before it knows which remote host we are connecting to, `SO_REUSEADDR` makes it possible that the OS could give us a port number we are already using to talk to that host.  This would make our { bind_ip, bind_port, connect_ip, connect_port } tuple a repeat, and `connect()` will fail.  As far as I understand it, ending up in this scenario isn't the same situation as running out of sockets, I think the OS could return a repeat ephemeral port number at any time if we are using `SO_REUSEADDR` (but this could also vary by OS, not entirely sure).

If we drop the retry logic, I think we should just not use `SO_REUSEADDR`.  Since it seems rare that a user would get into a situation where `SO_REUSEADDR` was really necessary, I would be fine with removing that logic if you agree!

Reply:
Yeah I am fine with removing it. Users may not need it, and it is not a requirement as far as we know. We can start with the more straightforward implementation without retrying, and add complexity later if we discover a need.